pkg/controller/controller_test.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/latest"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/testclient"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// testControllerPlugin tests a Controller's ability to respond to watches on more than one object
+type testControllerPlugin struct {
+	host                     ControllerHost
+	pvSource                 *framework.FakeControllerSource
+	pvcSource                *framework.FakeControllerSource
+	expectedClaimAddedCount  int
+	expectedVolumeAddedCount int
+	actualClaimAddedCount    int
+	actualClaimDeletedCount  int
+	actualVolumeAddedCount   int
+	actualVolumeDeletedCount int
+	lock                     sync.Mutex
+	uids                     map[string]interface{}
+}
+
+func NewTestController(pvCount, pvcCount int) *testControllerPlugin {
+	return &testControllerPlugin{
+		expectedVolumeAddedCount: pvCount,
+		expectedClaimAddedCount:  pvCount,
+		actualClaimAddedCount:    0,
+		actualClaimDeletedCount:  0,
+		actualVolumeAddedCount:   0,
+		actualVolumeDeletedCount: 0,
+		uids: make(map[string]interface{}),
+	}
+}
+
+func (plugin *testControllerPlugin) Init(host ControllerHost) {
+	plugin.host = host
+}
+
+func (plugin *testControllerPlugin) Name() string {
+	return "test-controller-plugin"
+}
+
+func (plugin *testControllerPlugin) DefaultSyncPeriod() time.Duration {
+	return 1 * time.Second
+}
+
+func (plugin *testControllerPlugin) CreateController() Controller {
+	return plugin
+}
+
+func (plugin *testControllerPlugin) Configs() []*framework.Config {
+	// source simulates an apiserver object endpoint.
+	plugin.pvSource = framework.NewFakeControllerSource()
```

Comment:
This triggered my comment above on the semantics of this method.  I expected a function called Configs() to just return me the cached configs and be idempotent, which this is *not*.